File: 3387-maximize-amount-after-two-days-of-conversions/3387-maximize-amount-after-two-days-of-conversions.java
	class Solution {
		public double maxAmount(String initialCurrency, List<List<String>> pairs1, double[] rates1, List<List<String>> pairs2, double[] rates2) {
			Map<String, Double> has = new HashMap<>();
			has.put(initialCurrency, 1.0);
			{
				Queue<String> q = new ArrayDeque<>();
				q.add(initialCurrency);
				while (!q.isEmpty()) {
					String cur = q.poll();
					for (int i = 0; i < pairs1.size(); i++) {
						List<String> pair = pairs1.get(i);
						if (pair.get(0).equals(cur)) {
							String next = pair.get(1);
							double rate = rates1[i];
							if (!has.containsKey(next)) {
								has.put(next, has.get(cur) * rate);
								q.add(next);
							}
						} else if (pair.get(1).equals(cur)) {
							String next = pair.get(0);
							double rate = 1.0 / rates1[i];
							if (!has.containsKey(next)) {
								has.put(next, has.get(cur) * rate);
								q.add(next);
							}
						}
					}
				}
			}

			double max = 1;
			for(String k : has.keySet()) {
				Map<String, Double> nex = new HashMap<>();
				nex.put(k, has.get(k));
				Queue<String> q = new ArrayDeque<>();
				q.add(k);
				while(!q.isEmpty()){
					String cur = q.poll();
					for(int i = 0;i < pairs2.size();i++){
						List<String> pair = pairs2.get(i);
						if(pair.get(0).equals(cur)){
							String next = pair.get(1);
							double rate = rates2[i];
							if(!nex.containsKey(next)){
								nex.put(next, nex.get(cur) * rate);
								q.add(next);
							}
						}else if(pair.get(1).equals(cur)){
							String next = pair.get(0);
							double rate = 1.0 / rates2[i];
							if(!nex.containsKey(next)){
								nex.put(next, nex.get(cur) * rate);
								q.add(next);
							}
						}
					}
				}
				max = Math.max(max, nex.getOrDefault(initialCurrency, 0.0));
			}
			return max;
		}
	}

File: 3387-maximize-amount-after-two-days-of-conversions/3387-maximize-amount-after-two-days-of-conversions.cpp
class Solution {
public:
    map<string, double> calc(string s0, vector<vector<string>>& pairs, vector<double>& rates) {
        map<string, vector<pair<string, double>>> e;
        int n = pairs.size();
        for (int i = 0; i < n; i++) {
            string s1 = pairs[i][0], s2 = pairs[i][1];
            e[s1].emplace_back(s2, rates[i]);
            e[s2].emplace_back(s1, 1 / rates[i]);
        }
        queue<string> q;
        map<string, double> ret;
        q.push(s0);
        ret[s0] = 1.0;
        while(!q.empty()) {
            string now = q.front();
            q.pop();
            for (auto &t : e[now]) {
                string nxt = t.first;
                double r = t.second;
                if (ret.contains(nxt)) continue;
                ret[nxt] = ret[now] * r;
                q.push(nxt);
            }
        }
        return ret;
    }
    double maxAmount(string s0, vector<vector<string>>& pairs1, vector<double>& rates1, vector<vector<string>>& pairs2, vector<double>& rates2) {
        for (auto &t : rates2) {
            t = 1 / t;
        }
        double ans = 1.0;
        auto m1 = calc(s0, pairs1, rates1);
        auto m2 = calc(s0, pairs2, rates2);
        for (auto &t : m1) {
            if (m2.contains(t.first)) {
                ans = max(ans, m1[t.first] * m2[t.first]);
            }
        }
        return ans;
    }
};